generator/sbpg/targets/rust.py

```diff
@@ -101,11 +128,37 @@ def parse_type(field):
     # This is an inner class, call default constructor
     return "%s::parse(_buf)" % field.type_id
 
+def gps_time(msg, all_messages):
+  def time_aware_header(type_id):
+    for m in all_messages:
+      if m.identifier == type_id:
+        return any([f.identifier == "t" for f in m.fields])
```

Comment:
Talking with @john-michaelburke we worked out that MSG_OBS and MSG_OSR are a good example where this heuristic doesn't work (for ordering purposes) -- the .t field in the ObservationHeader type is the time of the observation/correction, which could be several seconds before the correct GPS time.  In the console (and the SBP chopper) these messages should be considered part of the GNSS epoch that they are closest too instead of the time that's indicated on the messages.

Reply:
Also looking at the choppa more there seems to be a few more edge cases

- https://github.com/swift-nav/thachoppa/blob/af018c8922749ca7db9a1c1306e0d339babd1bab/src/lib.rs#L135
- https://github.com/swift-nav/thachoppa/blob/af018c8922749ca7db9a1c1306e0d339babd1bab/src/lib.rs#L87

Both of those messages we just treat as normal but it appears they have some special handling required.

Reply:
> these messages should be considered part of the GNSS epoch that they are closest too instead of the time that's indicated on the messages.

Could this be another category of messages? Like we have time-aware, meaning that message has a time useful for ordering. Then there are "contextually time aware" messages that on their own aren't useful but with additional information can be. Then everything else that's not time aware.

Reply:
The context is definitely very important for what we're building here, I think we need to describe a couple different things:
- any time from a message that maybe be useful
- messages that don't have a time, or that have a time that's not useful in any context we describe with the type system
- time from a message that's useful for ordering or slicing a stream of messages *originating from a device*

So, maybe this could be an enum of some sort:
```rust
enum MessageTime {
  Unknown,
  Base(GpsTime),
  Rover(GpsTime),
}
```

Then we could implement different `PartialOrd` traits depending on use case:
- one that only compares `MessageTime::Rover` for ordering-- that is, only `PartialOrd(Rover, Rover)` can produce an ordering, `PartialOrd(Rover, X)` or `PartialOrd(X, Rover)` should will produce `Ordering::Equal` -- the use case here would be the current use case of ordering and slicing messages from a rover
- one that does the same thing, but only compares `PartialOrd(Base, Base)` -- use case would be ordering/slicing messages from a base station or correction network
- one that compares either `PartialOrd(Base, Rover)`,  `PartialOrd(Rover, Base)`, `PartialOrd(Rover, Rover)`, or  `PartialOrd(Base, Base)` -- not sure what the use case would be since it compares things that are technically "incomparable", but could be included for completeness

Reply:
> 
> 
> Also looking at the choppa more there seems to be a few more edge cases
> 
>     * https://github.com/swift-nav/thachoppa/blob/af018c8922749ca7db9a1c1306e0d339babd1bab/src/lib.rs#L135
> 
>     * https://github.com/swift-nav/thachoppa/blob/af018c8922749ca7db9a1c1306e0d339babd1bab/src/lib.rs#L87
> 
> 
> Both of those messages we just treat as normal but it appears they have some special handling required.

So, under the idea described above we should "denylist" MSG_OBS, MSG_OSR (and MSG_SSR?) from the current ordering and classify it as `MessageTime::Base` and for the Imu message referenced in the thechoppa, we'd probably need a custom trait implementation pulled from theachoppa.

Reply:
Further expanding on the SBP chopper use case, we probably need to have a means to specify a "reference time" when generating a GpsTime for comparison, the sbp chopper this is done when updating the "running clock": https://github.com/swift-nav/thachoppa/blob/master/src/lib.rs#L166.  Many messages only include a TOW, so this would allow the generation of a GpsTime to fill in the `wn` field from a reference time.

Reply:
> we probably need to have a means to specify a "reference time" when generating a GpsTime for comparison

What if we changed the method to be -

```rust
fn gps_time(&self, reference: impl Into<Option<GpsTime>>) -> Option<Result<GpsTime>>
```

Or maybe we could have two different methods

Reply:
> > we probably need to have a means to specify a "reference time" when generating a GpsTime for comparison
> 
> What if we changed the method to be -
> 
> ```rust
> fn gps_time(&self, reference: impl Into<Option<GpsTime>>) -> Option<Result<GpsTime>>
> ```
> 
> Or maybe we could have two different methods

This sounds fine

Reply:
Also, I wonder if an iterator adapter could be a good solution here. Something to transform 

`Iterator<Item = SBP>` -> `Iterator<Item = Wrapper>`

```rust
enum Wrapper {
    Message(SBP),
    TimeChanged(GpsTime),
}
```

Whenever a message would change the current time, a `Wrapper::Time` is returned before that message. We could have one adapter for `MessageTime::Base` and one for `MessageTime::Rover`.  With an extension trait like 

```rust
trait SBPIterExt: Iterator<Item = SBP> {
    fn rover_time(self) -> RoverTimeIter;
    fn base_time(self) -> BaseTimeIter;
}
```
people can do `iter_messages(rdr).rover_time().etc()`. also could keep it simple and just expose the adapter it's like

```rust
let messages = iter_messages(rdr);
for _ in RoverTimeIter::new(messages) {}
```

Either way - the adapter seems useful for holding state like the running clock/reference time.